lib/ardb/has_slug.rb

```diff
@@ -0,0 +1,101 @@
+module Ardb
+
+  module HasSlug
+
+    def self.included(klass)
+      klass.class_eval do
+        extend ClassMethods
+        include InstanceMethods
+
+      end
+    end
+
+    module ClassMethods
+
+      def has_slug(options = nil)
+        options ||= {}
+        raise(ArgumentError, "a source must be provided") unless options[:source]
+
+        @ardb_has_slug_attribute         = options[:attribute] || :slug
+        @ardb_has_slug_source_proc       = options[:source].to_proc
+        @ardb_has_slug_preprocessor_proc = (options[:preprocessor] || :downcase).to_proc
+        @ardb_has_slug_allow_underscores = !!options[:allow_underscores]
+
+        # can't always b/c of slug is written `after_save`
+        validates_presence_of(self.has_slug_attribute, :on => :update)
+        validates_uniqueness_of(self.has_slug_attribute, {
+          :case_sensitive => true,
+          :scope          => options[:unique_scope]
+        })
+
+        after_create :ardb_generate_slug
+        after_update :ardb_generate_slug
+      end
+
+      def has_slug_attribute
+        @ardb_has_slug_attribute
+      end
+
+      def has_slug_source_proc
+        @ardb_has_slug_source_proc
+      end
+
+      def has_slug_preprocessor_proc
+        @ardb_has_slug_preprocessor_proc
+      end
+
+      def has_slug_allow_underscores
+        @ardb_has_slug_allow_underscores
+      end
```

Comment:
@jcredding really don't love adding all these class methods.  Other implementation didn't do this.